File: app/src/main/java/com/example/weizhenbin/wangebug/views/autoscrolllayout/AutoScrollerViewPager.java
package com.example.weizhenbin.wangebug.views.autoscrolllayout;

import android.content.Context;
import android.os.Handler;
import android.os.Message;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.view.PagerAdapter;
import android.support.v4.view.ViewPager;
import android.util.AttributeSet;
import android.util.Log;

import java.lang.ref.WeakReference;

/**
 * Created by weizhenbin on 2018/8/27.
 * 自动滚懂viewpgae
 */

public class AutoScrollerViewPager extends ViewPager {

    private static final int SCROLLER_KEY=100;

    private Handler handler;

    /**时间间隔 毫秒
     * */
    private int timeIntervalMs=3000;

    private boolean autoScrollAble=true;

    public AutoScrollerViewPager(@NonNull Context context) {
        this(context,null);
    }

    public AutoScrollerViewPager(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        initHandler();
    }

    public void setTimeIntervalMs(int timeIntervalMs) {
        this.timeIntervalMs = timeIntervalMs;
    }

    private static class ViewPageHanlder extends Handler{
        private WeakReference<AutoScrollerViewPager> weakReference;
        private PagerAdapter pagerAdapter;
        private int pageCount;
        private int currentIndex=0;
        private ViewPageHanlder(AutoScrollerViewPager autoScrollerViewPager) {
            this.weakReference = new WeakReference<AutoScrollerViewPager>(autoScrollerViewPager);
            pagerAdapter= weakReference.get().getAdapter();
        }

        @Override
        public void handleMessage(Message msg) {
            super.handleMessage(msg);
            if (msg.what==SCROLLER_KEY){
                Log.d("ViewPageHanlder", "pagerAdapter:" + pagerAdapter);
                if (pagerAdapter==null) {
                    pagerAdapter=weakReference.get().getAdapter();
                    if (pagerAdapter!=null) {
                        pageCount = pagerAdapter.getCount();
                    }
                }
                currentIndex=weakReference.get().getCurrentItem();
                if (currentIndex<pageCount) {
                    currentIndex++;
                }
                weakReference.get().setCurrentItem(currentIndex);
                sendEmptyMessageDelayed(SCROLLER_KEY,weakReference.get().timeIntervalMs);
                Log.d("ViewPageHanlder", "自动滚动:"+currentIndex+"pageCount:"+pageCount);
            }
        }
    }

    private void initHandler() {
        handler=new ViewPageHanlder(this);
    }


    @Override
    public  void setAdapter(@Nullable PagerAdapter adapter) {
        super.setAdapter(adapter);
    }

    @Override
    protected void onAttachedToWindow() {
        super.onAttachedToWindow();
        if (autoScrollAble) {
            handler.sendEmptyMessageDelayed(SCROLLER_KEY, timeIntervalMs);
        }
    }

    public void setAutoScrollAble(boolean autoScrollAble) {
        this.autoScrollAble = autoScrollAble;
    }

    @Override
    protected void onDetachedFromWindow() {
        super.onDetachedFromWindow();
        handler.removeCallbacksAndMessages(null);
    }
}

File: app/src/main/java/com/example/weizhenbin/wangebug/views/autoscrolllayout/AutoScrollerViewPager.kt
package com.example.weizhenbin.wangebug.views.autoscrolllayout

import android.content.Context
import android.os.Handler
import android.os.Message
import android.support.v4.view.PagerAdapter
import android.support.v4.view.ViewPager
import android.util.AttributeSet

import java.lang.ref.WeakReference

/**
 * Created by weizhenbin on 2018/8/27.
 * 自动滚懂viewpgae
 */

class AutoScrollerViewPager @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null) : ViewPager(context, attrs) {

    private var myHandler: Handler? = null

    /**时间间隔 毫秒
     */
    private var timeIntervalMs = 3000

    private var autoScrollAble = true

    init {
        initHandler()
    }

    fun setTimeIntervalMs(timeIntervalMs: Int) {
        this.timeIntervalMs = timeIntervalMs
    }

    private class ViewPageHandle  constructor(autoScrollerViewPager: AutoScrollerViewPager) : Handler() {
        private val weakReference: WeakReference<AutoScrollerViewPager> = WeakReference(autoScrollerViewPager)
        private var pagerAdapter: PagerAdapter? = null
        private var pageCount: Int = 0
        private var currentIndex = 0

        init {
            weakReference.get()?.let {
                pagerAdapter = it.adapter
            }
        }

        override fun handleMessage(msg: Message) {
            super.handleMessage(msg)
            weakReference.get()?.let {
                if (msg.what == SCROLLER_KEY) {
                    if (pagerAdapter == null) {
                        pagerAdapter = it.adapter
                        if (pagerAdapter != null) {
                            pageCount = pagerAdapter!!.count
                        }
                    }
                    currentIndex = it.currentItem
                    if (currentIndex < pageCount) {
                        currentIndex++
                    }
                    it.currentItem = currentIndex
                    sendEmptyMessageDelayed(SCROLLER_KEY, it.timeIntervalMs.toLong())
                }
            }

        }
    }

    private fun initHandler() {
        myHandler = ViewPageHandle(this)
    }


    override fun setAdapter(adapter: PagerAdapter?) {
        super.setAdapter(adapter)
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        if (autoScrollAble) {
            myHandler!!.sendEmptyMessageDelayed(SCROLLER_KEY, timeIntervalMs.toLong())
        }
    }

    fun setAutoScrollAble(autoScrollAble: Boolean) {
        this.autoScrollAble = autoScrollAble
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        myHandler!!.removeCallbacksAndMessages(null)
    }

    companion object {

        private val SCROLLER_KEY = 100
    }
}
